### PR TITLE
WI #2509 Add intrinsic function from Cobol 5.1 to 6.3

### LINKED ISSUE #2509: Add intrinsic function from Cobol 5.1 to 6.3
**What is the current problem**
Recognize  intrinsic `function` from Cobol 5.1 to 6.3 but without checking arguments.
`function` keyword is still mandatory in this issue.
Repository functionnality / optional `function` keyword is for another issue.

Functions to add:
- function ULENGTH
- function UPOS
- function USUBSTR
- function UVALID
- function UWIDTH
- function E
- function PI
- function TRIM
- function ABS
- function BYTE-LENGTH
- function EXP
- function EXP10
- function NUMVAL-F
- function SIGN
- function TEST-NUMVAL
- function TEST-NUMVAL-C
- function TEST-NUMVAL-F
- function BIT-OF
- function HEX-OF
- function BIT-TO-CHAR
- function HEX-TO-CHAR
- function COMBINED-DATETIME
- function FORMATTED-CURRENT-DATE
- function FORMATTED-DATE
- function FORMATTED-DATETIME
- function FORMATTED-TIME
- function INTEGER-OF-FORMATTED-DATE
- function SECONDS-FROM-FORMATTED-TIME
- function SECONDS-PAST-MIDNIGHT
- function TEST-DATE-YYYYMMDD
- function TEST-DAY-YYYYDDD
- function TEST-FORMATTED-DATETIME
- function UUID4

TODO check if something is missing

Cobol code to illustrate.
```cobol
TODO
```

**Technical**
Will be edited by a team member.
Implementation details with the chosen solution.

**How to test automatically**
Will be edited by a team member.
Only if our current automated tests doesn't handle this.

### PR DESCRIPTION
**Pull request by efr15**
Fixes #2509

A new intrinsic function (TRIM) also accepts keyword as argument => some changes in grammar

### CODE REVIEW

File: TypeCobol/Compiler/CodeElements/Expressions/CallParameter.cs [hunk 1]
```diff
@@ -160,4 +160,14 @@ public enum ParameterPassingDirection
         Output,
         Returning
     }
+
+    /// <summary>
+    /// A keyword shall be specified in accordance with the intrinsic function definition.
+    /// For instance the keywords LEADING and TRAILING may be specified as the second and optional argument of TRIM.
+    /// </summary>
+    public enum IntrinsicFunctionKeywordArgument
```
Comment (fm-117) [inline TypeCobol/Compiler/CodeElements/Expressions/CallParameter.cs:168]: I'm not ok with the naming. `IntrinsicFunctionKeywordArgument` is related to the technical aspect of our parser. I'd rather go with `TrimDirection`

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
 using TypeCobol.Compiler.Diagnostics;
 using TypeCobol.Compiler.Scanner;
+using static TypeCobol.Compiler.Parser.CobolWordsBuilder;
```
Comment (fm-117) [inline TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs:8]: required ?

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs [hunk 3]
```diff
@@ -311,7 +312,42 @@ internal StorageArea CreateFunctionIdentifier(CodeElementsParser.FunctionIdentif
         [NotNull]
         internal FunctionCall CreateIntrinsicFunctionCall(CodeElementsParser.IntrinsicFunctionCallContext context) {
             var name = CobolWordsBuilder.CreateIntrinsicFunctionName(context.IntrinsicFunctionName());
-            return new IntrinsicFunctionCall(name, CreateArguments(context.argument()));
```
Comment (fm-117) [inline TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs:314]: The previous method `CreateArguments` should not be called anymore right ?

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs [hunk 4]
```diff
@@ -311,7 +312,42 @@ internal StorageArea CreateFunctionIdentifier(CodeElementsParser.FunctionIdentif
...
+                else if (intrinsicArgumentContext[i].LEADING() != null)
+                {
+                    Token token = ParseTreeUtils.GetFirstToken(intrinsicArgumentContext[i].LEADING());
+                    storageAreaOrValue = new Variable(new EnumeratedValue(token, typeof(IntrinsicFunctionKeywordArgument)));
+                }
+                else if (intrinsicArgumentContext[i].TRAILING != null)
+                {
+                    Token token = ParseTreeUtils.GetFirstToken(intrinsicArgumentContext[i].TRAILING());
+                    storageAreaOrValue = new Variable(new EnumeratedValue(token, typeof(IntrinsicFunctionKeywordArgument)));
+                }
```
Comment (fm-117) [inline TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs:342]: Could you factorize the part inside the if ?

```csharp
else
{
    var trimDirection = intrinsicArgumentContext[i].LEADING() ?? intrinsicArgumentContext[i].TRAILING();
    if (trimDirection != null)
    {
        Token token = ParseTreeUtils.GetFirstToken(trimDirection);
        storageAreaOrValue = new Variable(new EnumeratedValue(token, typeof(IntrinsicFunctionKeywordArgument)));
    }
}
```

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs (hunk 3 above)
Comment (efr15) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs:314]: The method `CreateArguments ` is still needed by `CreateUserDefinedFunctionCall`.

Comment (fm-117) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs:314]: Ok then there should be some factorization to do between the two.

Comment (efr15) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs:314]: Done

File: TypeCobol/Compiler/CodeElements/Expressions/CallParameter.cs (hunk 1 above)
Comment (efr15) [reply TypeCobol/Compiler/CodeElements/Expressions/CallParameter.cs:168]: As discussed, we keep it as it is (those terms are used by IBM).
But I've added a comment on the enumeration items to point out that the keywords are currently only allowed for TRIM.

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs [hunk 5]
```diff
@@ -311,7 +311,43 @@ internal StorageArea CreateFunctionIdentifier(CodeElementsParser.FunctionIdentif
...
+                        Variable enumeratedValue = new Variable(new EnumeratedValue(token, typeof(IntrinsicFunctionKeywordArgument)));
+                        arguments[i] = new CallSiteParameter() { StorageAreaOrValue = enumeratedValue };
+                    }
+                }
+            }
+            _insideFunctionArgument = false;
+            return arguments;
+        }
+
+        private CallSiteParameter CreateArgumentWithSharedVariableOrExpression(CodeElementsParser.SharedVariableOrExpression1Context sharedVariableOrExpression1Context)
```
[the remark this answers is not in the document]
Comment (efr15) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs]: Done

File: TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs [hunk 6]
```diff
@@ -311,7 +311,43 @@ internal StorageArea CreateFunctionIdentifier(CodeElementsParser.FunctionIdentif
...
+            CallSiteParameter[] arguments = new CallSiteParameter[intrinsicArgumentContext.Length];
+            for (int i = 0; i < intrinsicArgumentContext.Length; i++)
+            {
+                if (intrinsicArgumentContext[i].argument() != null)
+                {
+                    arguments[i] = CreateArgumentWithSharedVariableOrExpression(intrinsicArgumentContext[i].argument().sharedVariableOrExpression1());
+                }
+                else
+                {
+                    var keyword = intrinsicArgumentContext[i].LEADING() ?? intrinsicArgumentContext[i].TRAILING();
```
[the remark this answers is not in the document]
Comment (efr15) [reply TypeCobol/Compiler/Parser/CodeElementBuilder/CobolExpressionsBuilder.cs]: Done

**Review by fm-117 — APPROVED** [review]

**Review by fm-117 — APPROVED** [review]

**Review by smedilol — APPROVED** [review]

**Pull request merged**